searchcore/src/vespa/searchcore/fdispatch/search/mergehits.cpp

```diff
@@ -89,28 +97,40 @@ FastS_InternalMergeHits(FastS_HitMerger<T> *merger)
         sortItr = sortRef;
     }
 
+    GlobalIdHasher seenGids;
+
     FastS_make_heap(heap, heapSize, FastS_MergeCompare<T, F>);
 
     while (pt < end) {
         node = *heap;
         FastS_assert(heapSize > 0);
+        bool useHit = seenGids.insert(node->NT_GetHit()->HT_GetGlobalID());
         if (F::UseSortData()) {
-            if (!F::DropSortData()) {
+            if (!F::DropSortData() && useHit) {
                 FastS_MergeCopySortData<T>(node, sortItr++, sortDataLen);
             }
             node->NT_GetSortDataIterator()->Next();
         }
-        FastS_MergeCopyHit<T>(node->NT_GetHit(), pt++);
+        if (useHit) {
+            FastS_MergeCopyHit<T>(node->NT_GetHit(), pt++);
+        }
         node->NT_NextHit();
         if (node->NT_GetNumHitsLeft() > 0) {
             FastS_pop_push_heap(heap, heapSize, node, FastS_MergeCompare<T, F>);
         } else {
             FastS_pop_heap(heap, heapSize--, FastS_MergeCompare<T, F>);
+            if (heapSize == 0) {
```

Comment:
please combine this with the existing stop criteria above:
while((pt < end) && (heapSIze > 0))